### PR TITLE
refactor(cli/js/io): Remove Deno.EOF, use null instead

### LINKED ISSUE #3932: Use null instead of Deno.EOF for read return types
Throughout `cli/js` and `std`, there are return types that look like `number | Deno.EOF`, `Uint8Array | Deno.EOF`, `ServerRequest | Deno.EOF`. These are a pain to use and force a proprietary type for no reason. With `null` we can use JS's null checking syntaxes to extract the important type.

### PR DESCRIPTION
**Pull request by nayeemrmn**
Closes #3932, towards #4933, cc @ry @bartlomieju

You can just treat this as a POC but it should be good to merge.

- I see `null` as the basic and idiomatic option here. It leaves much less to regret.
- We're more likely to inherit positive changes to the language like this. Example: relatively recent nullish coalescing is used in a few places in this PR.
- Using a proprietary symbol to sidestep bad JS habits with `number | null` is adding constructs to solve non-Deno-specific (rather JS-specific) issues which I think we should avoid.
- Consider what would be chosen on the web platform if an applicable API was added there.

### CODE REVIEW

**Comment by ry** [comment]
I’m in favor of thing - I think. I want to get some more opinions first before landing.

Thanks @nayeemrmn for putting this together - I tried the same thing a few days ago and got stuck.

**Comment by kitsonk** [comment]
I've been 👍 for using `null`.  Using a symbol has quite a lot of unintentional downstream effects on support spec compliant APIs, where we really should be interpreting the symbol and transforming it, instead of just leaking it like we have to this point.

The argument against is "it is nullish and falsey, which can cause issue in JavaScript", but that statement is damning of the whole language, and we can't fix that, and the symbol gets in the way.

Longer term, it _might_ be worth considering, especially for the async APIs, to follow the pattern of streams, where there are promises which provide signals for these things, as well as continuing to extend the iterable patterns, where `done` and `value` are always seperate, so no need to have special values.

File: cli/js/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -403,20 +398,19 @@ declare namespace Deno {
...
      *
-     * When `readSync()` encounters end-of-file condition, it returns `Deno.EOF`
-     * symbol.
+     * When `readSync()` encounters end-of-file condition, it returns `null`.
      *
      * When `readSync()` encounters an error, it throws with an error.
      *
      * Callers should always process the `n` > `0` bytes returned before
-     * considering the `EOF`. Doing so correctly handles I/O errors that happen
+     * considering the `null`. Doing so correctly handles I/O errors that happen
```
Comment (ry) [inline cli/js/lib.deno.ns.d.ts:406]: please write "considering the EOF (`null`)."

File: cli/js/lib.deno.ns.d.ts [hunk 2]
```diff
@@ -477,7 +471,7 @@ declare namespace Deno {
     seekSync(offset: number, whence: SeekMode): number;
   }
 
-  /** Copies from `src` to `dst` until either `EOF` is reached on `src` or an
+  /** Copies from `src` to `dst` until either `null` is read from `src` or an
```
Comment (ry) [inline cli/js/lib.deno.ns.d.ts:474]: Please adjust
```
until either EOF (`null`) is read...
```

File: cli/js/lib.deno.ns.d.ts [hunk 3]
```diff
@@ -611,8 +602,8 @@ declare namespace Deno {
 
   /** Synchronously read from a resource ID (`rid`) into an array buffer (`buffer`).
    *
-   * Returns either the number of bytes read during the operation or End Of File
-   * (`Symbol(EOF)`) if there was nothing to read.
+   * Returns either the number of bytes read during the operation or `null` if
+   * there was nothing to read.
```
Comment (ry) [inline cli/js/lib.deno.ns.d.ts:606]: EOF is still a concept, even if we aren't using a symbol. It's important that the documentation is specific here:


> Returns either the number of bytes read during the operation or End Of File (`null`) if there was nothing to read.

Comment (ry) [reply cli/js/lib.deno.ns.d.ts:606]: We should comment that it is possible for `read()` to return successfully with `0` bytes and that it does not indicate EOF.

File: cli/js/tests/net_test.ts [hunk 4]
```diff
@@ -523,7 +523,7 @@ unitTest(
       try {
         while (true) {
           const nread = await conn.read(p);
-          if (nread === Deno.EOF) {
+          if (nread == null) {
```
Comment (ry) [inline cli/js/tests/net_test.ts:526]: I checked that `null == 0` is false, so what you've done is correct. But it seems prudent to retain the triple equals here.

File: cli/js/io.ts [hunk 5]
```diff
@@ -84,7 +81,7 @@ export async function* iter(
   const b = new Uint8Array(bufSize);
   while (true) {
     const result = await r.read(b);
-    if (result === EOF) {
+    if (result == null) {
```
Comment (bartlomieju) [inline cli/js/io.ts:84]: Why are these checks using double equal (`==`) instead of triple?

File: cli/js/tests/net_test.ts (hunk 4 above)
Comment (nayeemrmn) [reply cli/js/tests/net_test.ts:526]: I personally just default to `==` unless the API has the possibility of two distinct values which require a strict equality check, which no good API should. And I like operators which treat nullish values as the same.

That said, I guess `===` is slightly safer for debugging so I'll make the change.

**Pull request merged**